// glou/lib/plugins/index.js
'use strict';

var path = require('path');

var _ = require('lodash');
var loadPlugins = require('gulp-load-plugins');

_.extend(
  module.exports,

  // load gulp-* plugins from glou
  loadPlugins({
    lazy: false,
    config: path.resolve(__dirname, '../../package.json'),
    scope: ['dependencies', 'peerDependencies']
  }),

  {
    loadPlugins: loadPlugins,

    // good duplexer based on 0.10 streams
    plexer: require('plexer'),

    // "merge" package uses 0.10 streams, and lets the user choose if parallel or series
    merge: require('merge2'),

    // do not use the package "multipipe" as it uses 0.8 streams
    multipipe: require('./multipipe'),

    // do not use the package "gulp-add-src" as it uses 0.8 streams
    addSrc: require('./add-src'),

    deleteStream: require('./delete-stream'),

    // do not use the package "gulp-util"
    noop: require('./noop'),

    swallow: require('./swallow'),

    sync: require('./sync'),

    start: require('./start'),

    devnull: require('./devnull'),

    restoreBase: require('./restore-base'),

    // do not use the package "gulp-util"
    log: require('./log'),
  }
);
